backend/src/schema/resolvers/follow.spec.js

```diff
@@ -40,84 +97,73 @@ afterEach(async () => {
+        authenticatedUser = null
+        const { errors, data } = await mutate({
+          mutation: mutationFollowUser,
+          variables,
+        })
+        expect(errors[0]).toHaveProperty('message', 'Not Authorised!')
+        expect(data).toMatchObject({ follow: null })
       })
     })
 
-    it('I can follow another user', async () => {
-      const res = await clientUser1.request(mutationFollowUser('u2'))
-      const expected = {
-        follow: true,
-      }
-      expect(res).toMatchObject(expected)
-
-      const { User } = await clientUser1.request(`{
-        User(id: "u2") {
-          followedBy { id }
-          followedByCurrentUser
-        }
-      }`)
-      const expected2 = {
-        followedBy: [{ id: 'u1' }],
+    test('I can follow another user', async () => {
+      const { data: result } = await mutate({
+        mutation: mutationFollowUser,
+        variables,
+      })
+      const expectedUser = {
+        name: user2.name,
+        followedBy: [{ id: user1.id, name: user1.name }],
         followedByCurrentUser: true,
       }
-      expect(User[0]).toMatchObject(expected2)
+      expect(result).toMatchObject({ follow: expectedUser })
     })
 
-    it('I can`t follow myself', async () => {
-      const res = await clientUser1.request(mutationFollowUser('u1'))
-      const expected = {
-        follow: false,
-      }
-      expect(res).toMatchObject(expected)
-
-      const { User } = await clientUser1.request(`{
-        User(id: "u1") {
-          followedBy { id }
-          followedByCurrentUser
-        }
-      }`)
-      const expected2 = {
+    test('I can`t follow myself', async () => {
+      variables.id = user1.id
+      const { data: result } = await mutate({ mutation: mutationFollowUser, variables })
+      const expectedResult = { follow: null }
+      expect(result).toMatchObject(expectedResult)
+
+      const { data } = await query({
+        query: userQuery,
+        variables: { id: user1.id },
+      })
+      const expectedUser = {
         followedBy: [],
         followedByCurrentUser: false,
       }
-      expect(User[0]).toMatchObject(expected2)
+      expect(data).toMatchObject({ User: [expectedUser] })
     })
   })
   describe('unfollow user', () => {
+    beforeEach(async () => {
+      variables = {
+        id: user2.id,
+        type: 'User',
+      }
+      await mutate({ mutation: mutationFollowUser, variables })
+    })
+
     describe('unauthenticated follow', () => {
-      it('throws authorization error', async () => {
-        // follow
-        await clientUser1.request(mutationFollowUser('u2'))
-        // unfollow
-        const client = new GraphQLClient(host)
-        await expect(client.request(mutationUnfollowUser('u2'))).rejects.toThrow('Not Authorised')
+      test('throws authorization error', async () => {
+        authenticatedUser = null
+        const { errors, data } = await mutate({ mutation: mutationUnfollowUser, variables })
```

Comment:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-18T22:08:13Z" title="Thursday, September 19th 2019, 12:08:13 am +02:00">Sep 19, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -40,84 +97,73 @@ afterEach(async () => {
 describe('follow', () => {
   describe('follow user', () => {
     describe('unauthenticated follow', () => {
-      it('throws authorization error', async () => {
-        const client = new GraphQLClient(host)
-        await expect(client.request(mutationFollowUser('u2'))).rejects.toThrow('Not Authorised')
+      test('throws authorization error', async () => {
+        authenticatedUser = null
+        const { errors, data } = await mutate({
+          mutation: mutationFollowUser,
+          variables,
+        })
+        expect(errors[0]).toHaveProperty('message', 'Not Authorised!')
+        expect(data).toMatchObject({ follow: null })
       })
     })
 
-    it('I can follow another user', async () => {
-      const res = await clientUser1.request(mutationFollowUser('u2'))
-      const expected = {
-        follow: true,
-      }
-      expect(res).toMatchObject(expected)
-
-      const { User } = await clientUser1.request(`{
-        User(id: "u2") {
-          followedBy { id }
-          followedByCurrentUser
-        }
-      }`)
-      const expected2 = {
-        followedBy: [{ id: 'u1' }],
+    test('I can follow another user', async () => {
+      const { data: result } = await mutate({
+        mutation: mutationFollowUser,
+        variables,
+      })
+      const expectedUser = {
+        name: user2.name,
+        followedBy: [{ id: user1.id, name: user1.name }],
         followedByCurrentUser: true,
       }
-      expect(User[0]).toMatchObject(expected2)
+      expect(result).toMatchObject({ follow: expectedUser })
     })
 
-    it('I can`t follow myself', async () => {
-      const res = await clientUser1.request(mutationFollowUser('u1'))
-      const expected = {
-        follow: false,
-      }
-      expect(res).toMatchObject(expected)
-
-      const { User } = await clientUser1.request(`{
-        User(id: "u1") {
-          followedBy { id }
-          followedByCurrentUser
-        }
-      }`)
-      const expected2 = {
+    test('I can`t follow myself', async () => {
+      variables.id = user1.id
+      const { data: result } = await mutate({ mutation: mutationFollowUser, variables })
+      const expectedResult = { follow: null }
+      expect(result).toMatchObject(expectedResult)
+
+      const { data } = await query({
+        query: userQuery,
+        variables: { id: user1.id },
+      })
+      const expectedUser = {
         followedBy: [],
         followedByCurrentUser: false,
       }
-      expect(User[0]).toMatchObject(expected2)
+      expect(data).toMatchObject({ User: [expectedUser] })
     })
   })
   describe('unfollow user', () => {
+    beforeEach(async () => {
+      variables = {
+        id: user2.id,
+        type: 'User',
+      }
+      await mutate({ mutation: mutationFollowUser, variables })
+    })
+
     describe('unauthenticated follow', () => {
-      it('throws authorization error', async () => {
-        // follow
-        await clientUser1.request(mutationFollowUser('u2'))
-        // unfollow
-        const client = new GraphQLClient(host)
-        await expect(client.request(mutationUnfollowUser('u2'))).rejects.toThrow('Not Authorised')
+      test('throws authorization error', async () => {
+        authenticatedUser = null
+        const { errors, data } = await mutate({ mutation: mutationUnfollowUser, variables })
```


You could also do this:
```js
await expect(mutate({ mutation: mutationUnfollowUser, variables })).resolves.toMatchObject({
  errors: [{'message': 'Not Authorised!'}],
  data: { unfollow: null },
})

Reply:
> [<img alt="vbelolapotkov" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/vbelolapotkov) **Authored by [vbelolapotkov](https://github.com/vbelolapotkov)**
_<time datetime="2019-09-19T07:16:15Z" title="Thursday, September 19th 2019, 9:16:15 am +02:00">Sep 19, 2019</time>_

---


Thanks, I'm aware of it but I personally prefer separating a call for the method under test and assertions for the result. However, I'm fine to change it if there is a pattern in the codebase, 
@roschaefer please, let me know if this is a case.

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-19T14:26:56Z" title="Thursday, September 19th 2019, 4:26:56 pm +02:00">Sep 19, 2019</time>_

---


It's the pattern I stick to right now, but totally up to you! I thought you would be interested to see this.

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-19T14:27:37Z" title="Thursday, September 19th 2019, 4:27:37 pm +02:00">Sep 19, 2019</time>_

---


I just fond `toMatchObject` quite robust for testing a graphql response with very helpful error messages.